File: 0778-swim-in-rising-water/0778-swim-in-rising-water.py
class Solution:
    def swimInWater(self, grid: List[List[int]]) -> int:
        n = len(grid)
        heap = [(grid[0][0], 0, 0)];
        visited = [[False]*n for _ in range (n)]
        dirs = [[1, 0], [-1,0],[0, 1], [0, -1]]
        while heap:
            height, x, y = heapq.heappop(heap)
            if visited[x][y]:
                continue
            visited[x][y] = True
            if x == n-1 and y == n-1:
                return height
            for i in range (4):
                newX = x + dirs[i][0]
                newY = y + dirs[i][1]
                if newX < 0 or newX >= n or newY <0 or newY >= n or visited[newX][newY]:
                    continue
                heapq.heappush(heap, (max(height, grid[newX][newY]), newX, newY))
        return 0;

File: 0778-swim-in-rising-water/0778-swim-in-rising-water.cpp
class Solution {
private:
    const vector<pair<int, int>> direction = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    vector<vector<pair<int, int>>> adj;
    vector<int> Distance;
    vector<bool> visited;

public:
    void Dijkstra(int source) {
        priority_queue<pair<int, int>, vector<pair<int, int>>,
                       greater<pair<int, int>>>
            heap;
        heap.push({Distance[source], source});
        while (!heap.empty()) {
            auto cur = heap.top();
            heap.pop();
            if (visited[cur.second])
                continue;
            visited[cur.second] = true;
            for (auto v : adj[cur.second]) {
                if (!visited[v.first] &&
                    Distance[v.first] > max(Distance[cur.second], v.second)) {
                    Distance[v.first] = max(Distance[cur.second], v.second);
                    heap.push({Distance[v.first], v.first});
                }
            }
        }
    }
    int swimInWater(vector<vector<int>>& grid) {
        int n = grid.size();
        adj.resize(n * n, {});
        visited.resize(n * n, false);
        Distance.resize(n * n, INT_MAX);
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                int u = i * n + j;
                for (int k = 0; k < 4; k++) {
                    int newX = i + direction[k].first;
                    int newY = j + direction[k].second;
                    if (newX < 0 || newX >= n || newY < 0 || newY >= n)
                        continue;
                    int v = newX * n + newY;
                    adj[u].push_back({v, grid[newX][newY]});
                }
            }
        }
        Distance[0] = grid[0][0];
        Dijkstra(0);
        return Distance[n * n - 1];
    }
};